### PR TITLE
feat: rewrite static ES imports to dynamic import() calls (CPL-209)

### PR DESCRIPTION
**Pull request by GTC6244**
## Summary

Implements CPL-209: Replace Deno module imports. Instead of requiring ES module mode for static `import` statements, this adds a preprocessing step that:

1. **Scans** user code for static `import` statements before `async function main`
2. **Strips** them from the source
3. **Generates** equivalent dynamic `import()` calls inside the async IIFE wrapper

This keeps execution in **script mode** (`execute_script()`), while the dynamic `import()` calls go through the existing `CdnModuleLoader` — preserving all SHA-384 integrity verification, TOFU, caching, and URL allowlisting from CPL-206.

**Import patterns supported:** named (`{ a, b }`), default, namespace (`* as`), renamed (`as`), side-effect, multi-line, default+named combos, inline integrity hashes, full CDN URLs.

### How it works

```js
// User writes:
import { z } from "zod@3.22.4/+esm";
async function main(params) { z.string().parse("hello"); }

// Runtime generates:
(async () => {
  const { z } = await import("zod@3.22.4/+esm");
  async function main(params) { z.string().parse("hello"); }
  const params = {...};
  const data = await main(params);
  if (typeof data !== "undefined") { LitActions.setResponse({ response: data }); }
})();
```

### Files changed

- **`import_rewriter.rs`** (new) — Import parser, rewriter, and dynamic import generator with 22 unit tests
- **`runtime.rs`** — Integrates the rewriter into `execute_js()`. No-import path is unchanged.
- **`lib.rs`** — Module registration
- **`tests/it.rs`** — Integration tests for import rewriting

## Test Coverage

```
COVERAGE: 23/24 paths tested (96%)
  Code paths: 22/23 (96%)
  User flows: 1/1 (100%)
QUALITY: ★★★: 22  ★★: 1
```

Tests: 54 total (32 unit + 22 integration). All pass.

## Pre-Landing Review

1 informational issue found and auto-fixed: specifier JS-escaping hardened with `js_escape()`.

## Adversarial Review

Claude adversarial + Codex adversarial ran. Key findings (all informational, none blocking):
- Import inside comments/strings before `main` will be rewritten (acceptable for v1 — Lit Actions follow a strict import-then-main pattern)
- Cache integrity bypass for inline hashes (pre-existing in CdnModuleLoader, not introduced here)
- Serial `await` creates cold-cache latency (covered by existing 15-min timeout)

## Test plan
- [x] All 32 unit tests pass (import_rewriter + cdn_module_loader)
- [x] All 22 integration tests pass (0 failures, 2 ignored)
- [x] No-import code path unchanged (verified by regression test)
- [x] Pre-landing review: 1 issue auto-fixed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: lit-actions/server/import_rewriter.rs [hunk 1]
```diff
@@ -0,0 +1,654 @@
...
+    pub imports: Vec<ParsedImport>,
+}
+
+/// Scan user code for static `import` statements before `async function main`,
+/// remove them, and return the parsed imports.
+///
+/// Code after `async function main` (and any non-import code before it) is
+/// preserved unchanged.
+pub(crate) fn rewrite_imports(code: &str) -> RewriteResult {
+    let main_pos = code.find("async function main").unwrap_or(code.len());
```
[the remark this answers is not in the document]
Comment (GTC6244) [reply lit-actions/server/import_rewriter.rs]: Fixed in 0cca322. Added `find_main_declaration()` that searches for `async function main` and verifies the next character is not an identifier continuation (e.g. rejects `main2`, `mainHelper`). Kept it simpler than the suggested full state-machine approach since the scanner already has comment/string awareness for the import scan region — the boundary check is the key missing piece.

File: lit-actions/server/import_rewriter.rs [hunk 2]
```diff
@@ -0,0 +1,654 @@
...
+        }
+    }
+
+    /// Parse a quoted string (single or double) and return its contents.
+    fn parse_string(&mut self) -> Option<String> {
+        let r = self.remaining();
+        let quote = *r.as_bytes().first()?;
+        if quote != b'"' && quote != b'\'' {
+            return None;
+        }
```
[the remark this answers is not in the document]
Comment (GTC6244) [reply lit-actions/server/import_rewriter.rs:477]: Fixed in 0cca322. Rewrote `parse_string` to walk bytes and skip `\\\\<char>` escape sequences, matching both the Copilot and Graphite suggestions. Used the simpler Graphite-style approach (skip escaped bytes without interpreting them) since we need the raw specifier string for the module loader, not a decoded value.

File: lit-actions/server/import_rewriter.rs [hunk 3]
```diff
@@ -0,0 +1,654 @@
...
+                "const {{ {} }} = await import(\"{spec}\");\n",
+                parts.join(", "),
+            ));
+        }
+
+        if has_namespace {
+            out.push_str(&format!("const {ns_name} = await import(\"{spec}\");\n",));
+        }
+    }
+    out
```
[the remark this answers is not in the document]
Comment (GTC6244) [reply lit-actions/server/import_rewriter.rs:289]: Fixed in 0cca322. When both namespace and named/default bindings exist (`import Default, * as NS from "..."`), we now import once into the namespace variable and destructure from it:

```js
const NS = await import("...");
const { default: Default } = NS;
```

Single `await import()` call instead of two.

File: lit-actions/server/import_rewriter.rs [hunk 4]
```diff
@@ -0,0 +1,654 @@
...
+    }
+
+    /// Parse a quoted string (single or double) and return its contents.
+    fn parse_string(&mut self) -> Option<String> {
+        let r = self.remaining();
+        let quote = *r.as_bytes().first()?;
+        if quote != b'"' && quote != b'\'' {
+            return None;
+        }
+        let end = r[1..].find(quote as char)? + 1;
```
[the remark this answers is not in the document]
Comment (GTC6244) [reply lit-actions/server/import_rewriter.rs]: Fixed in 0cca322. Same fix as the Copilot comment above — `parse_string` now walks bytes and skips `\\\\<char>` escape sequences so escaped quotes don't terminate the parse early. Thanks for the clear suggested fix.

File: lit-actions/server/import_rewriter.rs [hunk 5]
```diff
@@ -0,0 +1,847 @@
...
+
+/// Scan user code for static `import` statements before `async function main`,
+/// remove them, and return the parsed imports.
+///
+/// Code after `async function main` (and any non-import code before it) is
+/// preserved unchanged. The scanner is aware of single-line comments (`//`),
+/// block comments (`/* */`), and string/template literals so that import-like
+/// text inside those constructs is never mistakenly rewritten.
+pub(crate) fn rewrite_imports(code: &str) -> RewriteResult {
+    let main_pos = find_main_declaration(code);
```
[the remark this answers is not in the document]
Comment (GTC6244) [reply lit-actions/server/import_rewriter.rs:46]: Fixed in 03a133b. `find_main_declaration` now uses the same comment/string/template-literal skipping logic as the import scanner. Occurrences of `async function main` inside `//` comments, `/* */` blocks, strings, or backtick templates are skipped. Added two tests covering the `// async function main` and `"async function main"` edge cases.
